euphonic/design/06_brille_integration.md

```diff
@@ -0,0 +1,199 @@
+
+### Use case 1: Brille 'Force Constants' for Faster Interpolation
+
+This would involve adding another object to Euphonic, for now I'll call it
+`BrilleForceConstants` (although a better name might be needed) which would act
+like `ForceConstants`, but would perform linear instead of Fourier
+interpolation. To make the most of this the current `ForceConstants` in
+Euphonic might have to be updated to enable direct calculation of structure
+factors, debye waller etc. from `ForceConstants` (without intermediate
+eigenvectors), although this might not be required if calculations are
+appropriately chunked. Another update to make this integration easier would be
+adding symmetry information to Euphonic's `Crystal` object (so that Brille
+doesn't need to find it itself).
+
+**Example Syntax**
+
+From Python:
+
+```
+from euphonic import ureg, ForceConstants
+from euphonic.util import mp_grid
+
+fc = ForceConstants.from_castep('quartz.castep_bin')
+
+# Create BrilleForceConstants object
+interpolate_kwargs = {'use_c': True, 'n_threads': 4, 'asr': 'reciprocal'}
+bfc = fc.calculate_brille_force_constants(
+    grid='mesh', max_points=10000, **interpolate_kwargs)
+
+# Create DebyeWaller
+dw_phonons = bfc.calculate_qpoint_phonon_modes(mp_grid([6,6,6]))
+dw = dw_phonons.calculate_debye_waller(temperature=300*ureg('K'))
+
+# Calculate StructureFactor
+phonons = bfc.calculate_qpoint_phonon_modes(qpts, n_threads=4)
+sf = phonons.calculate_structure_factor(dw)
+```
+
+From Matlab:
+
+The question here is whether the interface should be 'Horace-like' with bespoke
+functions to make it easier to simulate cuts, or whether it should more closely
+mirror Euphonic's Python interface. The following example is somewhere
+inbetween and assumes a Matlab wrapper has been written (named 'meuphonic').
+
+```
+% Create ForceConstants and BrilleForceConstants using Euphonic-like syntax
+fc = meuphonic.ForceConstants.from_castep('quartz.castep_bin');
+kwargs = {'use_c', true, 'n_threads', 4};
+bfc = fc.calculate_brille_force_constants( ...
+    'grid', 'mesh', 'max_points', 10000, 'interpolate_kwargs', kwargs);
+
+fit_params = [1.0]; % temperature, scale
+pars = {fit_params,
+        'force_constants', bfc,  % accepts fc or bfc
+        'interpolate_kwargs', {'threads', 4}, % different depending on fc or bfc
+        'debye_waller_grid', [6,6,6],
+        'temperature', 300.0,
+        'conversion_mat', [1,0,0; 0,1,0; 0,0,-1], % Convert Horace Q -> Euphonic Q
+        'chunk', 10000};
+sim_cut = disp2sqw_eval(exp_cut, @meuphonic.horace_sqw, pars, fwhh);
+```
+
+**Advantages:**
+ - Would benefit both Horace and Euphonic users
+ - Allows Horace users to skip a step depending on if they want to use brille
+   or not
+
+**Disadvantages:**
+ - Would require additions to Euphonic's ForceConstants, and quite a few
+   changes to Brille's existing repos before this option is feasible
+ - The interface could end up being quite different to SpinW (does this
+   matter?)
```

Comment:
Just a comment on the current SpinW integration. For using SpinW through Brille, the syntax:
```matlab
tri = sw_model('triAF', 1); 
spinw_pars = {'mat', {'J1'}, 'hermit', false, 'formfact', true, 'resfun', 'gauss'}; 
brl_pars = {'max_volume', 1e-5};
brl = brillem.Brille(tri, brl_pars{:});
sim_cut = disp2sqw_eval(exp_cut, @brl.horace_sqw, {[J1val] spinw_pars{:}}, fwhm);
```
is currently useable. Brille specific options are passed to the constructor `brillem.Brille` and only SpinW related parameters are passed to `disp2sqw_eval`.

For using SpinW directly from Horace, the syntax is:
```matlab
tri = sw_model('triAF', 1); 
spinw_pars = {'mat', {'J1'}, 'hermit', false, 'formfact', true, 'resfun', 'gauss'}; 
sim_cut = disp2sqw_eval(exp_cut, @tri.horace_sqw, {[J1val] spinw_pars{:}}, fwhm);
```
E.g. the same as with Brille but with the two Brille related lines removed, and calling the `horace_sqw` method of the SpinW object directly. (`tri = sw_model('triAF', 1);` is just a convenient shortcut to create a simple spin wave model. Usually, users would need several lines to create a model themselves.)

The final use case is for users wanting to using Brille with SpinW without using Horace (calculating the linearly interpolated spin wave spectrum without reference to a Horace workspace). The syntax here is:
```matlab
tri = sw_model('triAF', 1); 
spinw_pars = {'hermit', false, 'formfact', true, 'resfun', 'gauss'}; 
brl_pars = {'max_volume', 1e-5};
brl = brillem.Brille(tri, brl_pars{:});
brl.fill(spinw_pars{:})
hkl_path = {[0 0 0] [1 1 1] [0 0 1] [0 0 0] 100}; 
spec = brl.spinwave(hkl_path);
sw_plotspec(sw_egrid(sw_neutron(spec), 'Evect', linspace(0,20,100)));
```
compared with the syntax without Brille:
```matlab
tri = sw_model('triAF', 1); 
spinw_pars = {'hermit', false, 'formfact', true, 'resfun', 'gauss'}; 
hkl_path = {[0 0 0] [1 1 1] [0 0 1] [0 0 0] 100}; 
spec = tri.spinwave(hkl_path, spinw_pars{:});
sw_plotspec(sw_egrid(sw_neutron(spec), 'Evect', linspace(0,20,100)));
```
The main difference is that the SpinW parameters has to be passed to the `fill` call which then passes all its arguments to the underlying calculator (e.g. SpinW). No parameters except the hkl path or grid should be passed to the Brillem `spinwave` method unlike for the pure-SpinW case where the SpinW parameters are passed there.

Reply:
The current Brille/SpinW syntax is very similar to the proposed Euphonic syntax so I think this is fine. The main difference between `brillem.Brille` (the standalone Brille Matlab interface) and `brillem.Euphonic` (the Brille+Euphonic Matlab interface) is that `brillem.Euphonic` is just a light wrapper for `brilleu` (a Python module interfacing Brille to Euphonic) and so does not have a many methods to interact with the Brille object. This is pertinent in the separation of keyword arguments or input parameters meant for Brille or for the calculator (Euphonic or SpinW). In the case of `brillem.Euphonic` both sets should be passed to the constructor at the same time; For `brillem.Brille` the constructor does not immediately construct the grid (e.g. invoke the calculator) so it only needs Brille specific arguments. Only secondary calls to `fill` (the `horace_sqw` method implicitly calls `fill` if a grid is not constructed or is out-of-date) need to be passed the arguments for the calculator (SpinW in the above examples). This separation could be useful or annoying depending on your point of view... (it could also make it harder to implement universal arguments which should apply to both Brille and the calculator).